compiler/noirc_frontend/src/hir_def/types.rs

```diff
@@ -134,7 +134,24 @@ pub enum Type {
 /// is expected (such as in an array length position) are expected to be of kind `Kind::Numeric`.
 #[derive(PartialEq, Eq, Clone, Hash, Debug, PartialOrd, Ord)]
 pub enum Kind {
+    /// Can bind to any type
+    Any,
```

Comment:
I believe that this or a roughly equivalent `Kind::Unresolved` is required for `_` types, e.g.
```noir
let x: Field = 3;
let z: _ = x;
```

And also unspecified types, i.e. a `UnresolvedTypeData::Unspecified` that's filled in later.

(In the `no implicit numeric generics` PR, I used `Option<Kind>` where the `None` case was used for these unspecified cases.)

However, I'm not sure how a `Kind::Unresolved` would be different. Perhaps we could throw an error if a `Kind::Any`/`Kind::Unresolved` is reached during monomorphization? Though that might be able to occur in valid code if a `_`-typed variable is completely unused.

Regardless, I'll comment and make an issue to follow up, since we'll likely want the metaprogramming builtin for fresh type variables to take an explicit `Kind` (or at least offer a variant that does).